### PR TITLE
Add GetCorrespondingLoop() (and Switch) branch operation extensions

### PR DESCRIPTION
**Pull request by wachulski**
Fixes dotnet/roslyn#28095

[jcouv update:] Summary from Fred is we're adding an API to get the corresponding switch or loop structure for a given `break` or other branch operation.

### CODE REVIEW

File: src/Compilers/Core/Portable/Operations/IBranchOperationExtensions.cs
Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/Operations/IBranchOperationExtensions.cs]: Same problem here.  A 'break' in a 'loop' in a 'switch' will end up returning the switch, even though the 'break' corresponds to the loop.

Comment (CyrusNajmabadi) [reply src/Compilers/Core/Portable/Operations/IBranchOperationExtensions.cs]: Yup.  GetCorrespondingLoop and GetCorrespondingSwitch sound like great names!  Thanks!

**Comment by wachulski** [comment]
Should I add Basic unit tests?

File: src/Compilers/Core/Portable/Operations/IBranchOperationExtensions.cs
Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/Operations/IBranchOperationExtensions.cs]: probably better to write this as a loop instead of with recursion.

File: src/Compilers/Core/Portable/Operations/IBranchOperationExtensions.cs
Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/Operations/IBranchOperationExtensions.cs]: this logic isn't quite right.  Consider if you had:  ```loop -> switch -> continue```.  The 'continue' corresponds to the 'loop'.  but because you'll hit the ISWitchOperation you'll end up returning null thanks to this check.

**Comment by 333fred** [comment]
> // Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.

Please just put these extension methods in the OperationExtensions class, in this same folder.

---
Refers to: src/Compilers/Core/Portable/Operations/IBranchOperationExtensions.cs:1 in c1ee137. [](commit_id = c1ee137019e9ae3c1446f5f7dfc97bc26eca124e, deletion_comment = False)

File: src/Compilers/Core/Portable/Operations/IBranchOperationExtensions.cs
Comment (333fred) [inline src/Compilers/Core/Portable/Operations/IBranchOperationExtensions.cs]: >t [](start = 42, length = 1)

Capitalization

Comment (333fred) [reply src/Compilers/Core/Portable/Operations/IBranchOperationExtensions.cs]: And elsewhere in the extensions.

---
In reply to: [200513121](https://github.com/dotnet/roslyn/pull/28285#discussion_r200513121) [](ancestors = 200513121)

File: src/Compilers/Core/Portable/Operations/IBranchOperationExtensions.cs
Comment (333fred) [inline src/Compilers/Core/Portable/Operations/IBranchOperationExtensions.cs]: Need to add the exception to the documentation.

File: src/Compilers/Core/Portable/Operations/IBranchOperationExtensions.cs
Comment (333fred) [inline src/Compilers/Core/Portable/Operations/IBranchOperationExtensions.cs]: We don't use raw strings in exception messages, as they cannot be localized. Please add this string to the `CodeAnalysisResources.resx`, and construct the message using `string.Format`. Note you'll have to make sure to run the build after you add the file to the resx to make sure that all the xlf files get updated (you should see ~13 or so `.xlf` files in the diff).

**Comment by 333fred** [comment]
> Should I add Basic unit tests?

Absolutely yes :)

---
In reply to: [402699535](https://github.com/dotnet/roslyn/pull/28285#issuecomment-402699535) [](ancestors = 402699535)

**Comment by 333fred** [comment]
Please add argument checking for `null` as well. Additionally, we should probably add checking to make sure that the given argument is _not_ in a CFG, as `ILoopOperation`s and `ISwitchOperation`s are removed from that. You can check this by casting the input parameter to an `Operation` and checking the `SemanticModel` property. For a CFG node, this will always be null.

**Comment by 333fred** [comment]
@wachulski if you have any issues responding to feedback, feel free to tag me and I'll take a look.

File: src/Compilers/Core/Portable/Operations/IBranchOperationExtensions.cs
Comment (wachulski) [reply src/Compilers/Core/Portable/Operations/IBranchOperationExtensions.cs]: I can bet I had found an example somewhere around and just followed it. But np, I'll correct.

File: src/Compilers/Core/Portable/Operations/IBranchOperationExtensions.cs
Comment (wachulski) [reply src/Compilers/Core/Portable/Operations/IBranchOperationExtensions.cs]: true, but I wanted to use `as` in there. Anyway - I'll need to correct as @CyrusNajmabadi hit me with a bug finding 😃

File: src/Compilers/Core/Portable/Operations/IBranchOperationExtensions.cs
Comment (wachulski) [reply src/Compilers/Core/Portable/Operations/IBranchOperationExtensions.cs]: sure

**Comment by wachulski** [comment]
> Should I add Basic unit tests?
> 
> Absolutely yes :)

I've asked as I haven't found that many of IOperation tests in Basic. I'll add them but I'm wondering where to put them.

**Comment by wachulski** [comment]
> Please add argument checking for null as well. Additionally, we should probably add checking to make sure that the given argument is not in a CFG, as ILoopOperations and ISwitchOperations are removed from that. You can check this by casting the input parameter to an Operation and checking the SemanticModel property. For a CFG node, this will always be null.

@333fred I'm a bit lost. `Operation` implements `SemanticModel` as explicit interface member implementation and offers `OwningSemanticModel` instead. Which one you meant?

```csharp
if (((Operation) branchOperation).OwningSemanticModel == null)
{
    throw new ArgumentException("The argument provided must not be CFG.");
}

if (branchOperation.SemanticModel == null)
{
    throw new ArgumentException("The argument provided must not be CFG.");
}
```

**Comment by 333fred** [comment]
@wachulski just a check for the `SemanticModel` property will work fine. An operation created as part of a CFG will not have it. Additionally, for wording, I'd suggest `The provided operation must not be part of a Control Flow Graph.`

> I've asked as I haven't found that many of IOperation tests in Basic. I'll add them but I'm wondering where to put them.

The IOperation VB tests are at `src/Compilers/VisualBasic/Test/IOperation/IOperation`.

**Comment by wachulski** [comment]
Done both: the check and VB tests.

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IBranchOperation.Extensions.cs
Comment (333fred) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IBranchOperation.Extensions.cs]: >RegularWithFlowAnalysisFeature [](start = 82, length = 30)

You shouldn't need flow analysis for this.

Comment (333fred) [reply src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IBranchOperation.Extensions.cs]: You should be fine with whatever the default `parseOptions` is, no need to pass an explicit parameter.

---
In reply to: [205600237](https://github.com/dotnet/roslyn/pull/28285#discussion_r205600237) [](ancestors = 205600237)

**Comment by 333fred** [comment]
Please add some tests to verify the exceptions of the API.

This is looking very good @wachulski, just a couple of small suggestions and we should be able to get this in.

**Comment by wachulski** [comment]
@333fred All right, the only test I'm still missing is the one which constructs `IBranchOperation` with empty semantic model. Could you help me with that? Particularly, for a given string source how to create a CFG referring to IBranchOperation with an empty semantic model?

**Comment by 333fred** [comment]
@wachulski looking into it, it looks we're asserting in all tests that `IBranchOperation` cannot be in a CFG. So no test is necessary for that. If you push the other tests, I'll take a look and get the other approvals you'll need.

**Comment by wachulski** [comment]
Just pushed tests

**Review by 333fred — APPROVED** [review]
LGTM (commit 9)

@dotnet/roslyn-compiler @dotnet/analyzer-ioperation, please review.

**Comment by 333fred** [comment]
Test failure is https://github.com/dotnet/roslyn/issues/28062

**Comment by 333fred** [comment]
@dotnet-bot retest windows_release_unit32_prtest please

File: src/Compilers/Core/Portable/PublicAPI.Unshipped.txt
Comment (jcouv) [inline src/Compilers/Core/Portable/PublicAPI.Unshipped.txt]: >GetCorrespondingLoop [](start = 61, length = 20)

Tagging @jaredpar @gafter for public API change. 

@333fred @mavasani I assume this API was agreed upon by the IOperations design team. Can you confirm?

**Comment by 333fred** [comment]
@wachulski, we're currently discussing on the IOperation design team whether we want to take this API. I'm hoping to resolve it today or tomorrow, but in the worst case we might not be able to review it until Wednesday afternoon. Thanks for your patience on this.

File: src/Compilers/Core/Portable/Operations/OperationExtensions.cs
Comment (mavasani) [inline src/Compilers/Core/Portable/Operations/OperationExtensions.cs]: Would the implementation be simpler if you walk up the parent chain and directly compare `IBranchOperation.Target` with `ILoopOperation.ExitLabel` or  `ILoopOperation.ContinueLabel` or `ISwitchOperation.ExitLabel`? At the minimum you should add an assert that for a non-null return value of `GetCorrespondingSwitch` and `GetCorrespondingLoop`, the label symbols match.

**Comment by 333fred** [comment]
@wachulski, feedback from the IOperation design meeting today:

We like the API, but we have a couple of changes. We'd like to combine these two separate APIs into a single `IOperation GetCorrespondingOperation(this IBranchOperation operation)`, that returns the loop or switch associated with the given branch. The user can then check the returned operation to see if it's null, a loop, or a switch, and cast appropriately as the analyzer requires.

**Comment by wachulski** [comment]
Done, I also refactored the walking up the parent as @mavasani suggested.

File: src/Compilers/Core/Portable/Operations/OperationExtensions.cs
Comment (mavasani) [inline src/Compilers/Core/Portable/Operations/OperationExtensions.cs]: I think we want to return `null` here instead of throwing. We can potentially extend this method in future to return the target [ICaseClauseOperation](http://source.roslyn.io/#Microsoft.CodeAnalysis/Operations/ICaseClauseOperation.cs,17) for [BranchKind.GoTo](http://source.roslyn.io/#Microsoft.CodeAnalysis/Operations/BranchKind.cs,28)

**Review by mavasani — APPROVED** [review]
LGTM, modulo replacing exception for `InvalidBranchKindForFindingCorrespondingOperation` with null return.

File: src/Compilers/Core/Portable/Operations/OperationExtensions.cs
Comment (333fred) [inline src/Compilers/Core/Portable/Operations/OperationExtensions.cs]: Nit: Would prefer the `||` on the other line, so the operation.Target.Equals lines up.

File: src/Compilers/Core/Portable/Operations/OperationExtensions.cs
Comment (gafter) [reply src/Compilers/Core/Portable/Operations/OperationExtensions.cs]: Agreed.

---
In reply to: [208969708](https://github.com/dotnet/roslyn/pull/28285#discussion_r208969708) [](ancestors = 208969708)

File: src/Compilers/Core/Portable/Operations/OperationExtensions.cs
Comment (gafter) [inline src/Compilers/Core/Portable/Operations/OperationExtensions.cs]: Please remove this.

File: src/Compilers/Core/Portable/Operations/OperationExtensions.cs
Comment (gafter) [inline src/Compilers/Core/Portable/Operations/OperationExtensions.cs]: >syntax [](start = 106, length = 6)

"(e.g. no loop or switch syntax, or the branch is not a break or continue)"

**Review by gafter — APPROVED** [review]
API change looks good except as noted.

File: src/Compilers/Core/Portable/CodeAnalysisResources.resx
Comment (gafter) [inline src/Compilers/Core/Portable/CodeAnalysisResources.resx]: >InvalidBranchKindForFindingCorrespondingOperation [](start = 14, length = 49)

I think we no longer need this.

**Review by 333fred — APPROVED** [review]

**Comment by 333fred** [comment]
Test failure was https://github.com/dotnet/roslyn/issues/18779

**Comment by 333fred** [comment]
@dotnet-bot retest windows_debug_vs-integration_prtest please

**Comment by 333fred** [comment]
@dotnet-bot retest windows_release_vs-integration_prtest please

**Comment by 333fred** [comment]
@jaredpar/@jinujoseph, should this be going to `master` or `features/compiler`, or some other branch?

**Comment by jinujoseph** [comment]
master sounds good to me , unless compiler has a reason not to

**Comment by 333fred** [comment]
@wachulski thanks for the contribution and your patience!

**Pull request merged**
